src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs

```diff
@@ -949,6 +950,13 @@ internal Task ValidateAndReconnect(Action beforeDisconnect, int timeout)
             return runningReconnect;
         }
 
+        // This is in its own method to avoid always allocating the lambda in ValidateAndReconnect
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private Task ValidateAndReconnectSetupReconnectContinuation(int timeout)
```

Comment:
How does refactoring Task.Run() into a separate method, improve allocations?

Reply:
The Task.Run call takes a lambda expression argument. In order to allow context capture a hidden type is created by the compiler and the parts of the context referenced inside the lambda body are packed into the new variable and passed in as state. Unfortunately this behaviour means that an instance is allocated sometimes even when it isn't going to be used because it won't be on the path taken through the method at runtime. 

By moving the call which contains the lambda declaration outside the method the hidden type is no longer on the stack frame for the calling method and isn't automatically allocated. It will be allocated in the extracted method and will contain on the parameters passed in.

If you look at the IL for any assembly which uses a lambda you'll be able to find a bunch of types called `<>cDisplayClass*` these are lambda capture machinery. When profiling a lot of these showed up in the allocation traces and since the synchronous paths never use those lambdas they were eligible to be made optional.

Reply:
"Pay for an allocation/assignments of a capture class regardless of whether it's used" https://github.com/dotnet/roslyn/issues/20777